kpi-tracker/src/main.rs

```diff
@@ -702,30 +765,131 @@ async fn use_node(db: &mut DB, from_height: AbsoluteBlockHeight) -> anyhow::Resu
+            let block_data = process_block(&mut node, block.block_hash).await?;
+            block_sender.send(BlockData::Normal(block_data)).await?;
             log::info!("Processed block ({}): {}", height, block.block_hash);
         }
     }
 
     Ok(())
 }
 
+/// Inserts the `block_data` collected for a single block into the database
+/// defined by `db`. Everything is commited as a single transactions allowing
+/// for easy restoration from the last recorded block (by height) inserted into
+/// the database.
+async fn db_insert_block(db: &mut DBConn, block_data: BlockData) -> anyhow::Result<()> {
+    let db_tx = db
+        .client
+        .transaction()
+        .await
+        .context("Failed to build DB transaction")?;
+
+    let tx_ref = &db_tx;
+    let prepared_ref = &db.prepared;
+
+    let insert_common = |block_hash: BlockHash, block_details: BlockDetails, accounts: Accounts| async move {
+        let block_id = prepared_ref
+            .insert_block(tx_ref, block_hash, block_details)
+            .await?;
+
+        for (address, details) in accounts.into_iter() {
+            prepared_ref
+                .insert_account(tx_ref, block_id, address, details)
+                .await?;
+        }
+
+        Ok::<_, DBError>(block_id)
+    };
+
+    match block_data {
+        BlockData::Genesis(GenesisBlockData {
+            block_hash,
+            block_details,
+            accounts,
+        }) => {
+            insert_common(block_hash, block_details, accounts).await?;
+        }
+        BlockData::Normal(NormalBlockData {
+            block_hash,
+            block_details,
+            accounts,
+            transactions,
+            contract_modules,
+            contract_instances,
+        }) => {
+            let block_id = insert_common(block_hash, block_details, accounts).await?;
+
+            for module_ref in contract_modules.into_iter() {
+                db.prepared
+                    .insert_contract_module(&db_tx, block_id, module_ref)
+                    .await?;
+            }
+
+            for (address, details) in contract_instances.into_iter() {
+                db.prepared
+                    .insert_contract_instance(&db_tx, block_id, address, details)
+                    .await?;
+            }
+
+            for (hash, details) in transactions.into_iter() {
+                db.prepared
+                    .insert_transaction(&db_tx, block_id, hash, details)
+                    .await?;
+            }
+        }
+    };
+
+    db_tx
+        .commit()
+        .await
+        .context("Failed to commit DB transaction.")?;
+    Ok(())
+}
+
+/// Runs a process of inserting data coming in on `block_receiver` in a database
+/// defined in [`Args.db_connection`]
+async fn run_db_process<'a>(
+    mut block_receiver: tokio::sync::mpsc::Receiver<BlockData>,
+    height_sender: tokio::sync::oneshot::Sender<AbsoluteBlockHeight>,
+) -> anyhow::Result<()> {
+    let args = Args::parse();
```

Comment:
Yeah, let's definitely **not** parse command line arguments in multiple places. That's why we have the configuration.

Reply:
Will change this, but can you elaborate on why it matters (out of curiosity?) I assume it's because it's more expensive to do multiple times and probably also bad practice maybe?

Reply:
Right, so the main reason is that it makes the data flow entirely unclear.

Programs should be structured so that there is the outer shell which gets user data and that is parsed into validated structures once. Then all the components that need the configuration should have that explicitly, e.g., via a struct.

The main advantage of this is that it makes it easier to understand programs and also maintain them, since it is a lot clearer where the different options are needed in the components of the program.

The other issue is also that we don't want to do redundant work, such as parsing command line arguments every time a function is called. It is also technically possible that a value of an environment variable changes during the running of the program, which would mean the behaviour of our program changes randomly based on external factors, which is not the best for stability.

Reply:
Thanks for the elaborative response!